### PR TITLE
Design DeleteAsset interface in Fybrik

### LINKED ISSUE #1242: Catalog Connector Delete Dataset
Add interface to delete a dataset from a data catalog , and implement an example implementation in katalog

### PR DESCRIPTION
**Pull request by rohithdv**
This PR add `DeleteAsset` method in the Fybrik Data catalog interface. The corresponding REST method is called `deleteAsset`. 

This PR partially fixes #1242 .

Signed-off-by: Rohith D Vallam [rovallam@in.ibm.com](mailto:rovallam@in.ibm.com)

### CODE REVIEW

File: charts/fybrik/files/taxonomy/datacatalog.json [hunk 1]
```diff
@@ -43,6 +43,34 @@
         }
       }
     },
+    "DeleteAssetRequest": {
+      "type": "object",
+      "required": [
+        "destinationCatalogID"
+      ],
```
Comment (shlomitk1) [inline charts/fybrik/files/taxonomy/datacatalog.json:50]: destinationAssetID is required as well

File: charts/fybrik/files/taxonomy/datacatalog.json [hunk 2]
```diff
@@ -43,6 +43,34 @@
...
+      "properties": {
+        "destinationAssetID": {
+          "description": "Asset ID of the to-be deleted asset",
+          "type": "string"
+        },
+        "destinationCatalogID": {
+          "description": "The destination catalog id in which the asset to be deleted exists",
```
Comment (shlomitk1) [inline charts/fybrik/files/taxonomy/datacatalog.json:57]: simplify: "... in which the asset exists"

File: charts/fybrik/files/taxonomy/datacatalog.json [hunk 3]
```diff
@@ -43,6 +43,34 @@
...
+      }
+    },
+    "DeleteAssetResponse": {
+      "type": "object",
+      "required": [
+        "status"
+      ],
+      "properties": {
+        "status": {
+          "description": "The ID of the created asset based on the source asset information given in CreateAssetRequest object",
```
Comment (shlomitk1) [inline charts/fybrik/files/taxonomy/datacatalog.json:69]: Not relevant to this request

File: charts/fybrik/files/taxonomy/datacatalog.json [hunk 4]
```diff
@@ -43,6 +43,34 @@
...
+        "destinationCatalogID": {
+          "description": "The destination catalog id in which the asset to be deleted exists",
+          "type": "string"
+        }
+      }
+    },
+    "DeleteAssetResponse": {
+      "type": "object",
+      "required": [
+        "status"
```
Comment (shlomitk1) [inline charts/fybrik/files/taxonomy/datacatalog.json:65]: Why would we need anything except for the returned code?

File: manager/controllers/mockup/datacatalog.go [hunk 5]
```diff
@@ -52,6 +52,13 @@ func (d *DataCatalogDummy) CreateAsset(in *datacatalog.CreateAssetRequest, creds
 	return &datacatalog.CreateAssetResponse{AssetID: "testAssetID"}, nil
 }
 
+func (m *DataCatalogDummy) DeleteAsset(in *datacatalog.DeleteAssetRequest, creds string) (*datacatalog.DeleteAssetResponse, error) {
+	// TODO
+	// will be provide a proper implementation once the implementation of DeleteAssetInfo in katalog-connector
+	// is completed in a future PR. Till then a dummy implementation is provided.
+	return &datacatalog.DeleteAssetResponse{Status: "DeleteAssetInfo not implemented in DataCatalogDummy"}, nil
+}
```
Comment (shlomitk1) [inline manager/controllers/mockup/datacatalog.go:60]: This is a mockup - not related to katalog-connector

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go [hunk 6]
```diff
@@ -13,6 +13,7 @@ import (
 	"emperror.dev/errors"
```
Comment (shlomitk1) [inline pkg/connectors/datacatalog/clients/datacatalog_grpc.go:13]: Not related to this PR. However, when will grpc support be removed?

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go [hunk 7]
```diff
@@ -211,3 +214,9 @@ func ConvertDataCatalogGrpcRespToOpenAPIResp(result *pb.CatalogDatasetInfo) (*da
 func (m *grpcDataCatalog) CreateAsset(in *datacatalog.CreateAssetRequest, creds string) (*datacatalog.CreateAssetResponse, error) {
 	return &datacatalog.CreateAssetResponse{AssetID: "testAssetID"}, nil
 }
+
+// just adding this dummy implementation as we are going to remove grpc support soon.
+// Then this file will be removed. Till then we provide a dummy implementation.
+func (m *grpcDataCatalog) DeleteAsset(in *datacatalog.DeleteAssetRequest, creds string) (*datacatalog.DeleteAssetResponse, error) {
```
Comment (shlomitk1) [inline pkg/connectors/datacatalog/clients/datacatalog_grpc.go:220]: There is no such functionality in the proto, so why is this needed?

File: pkg/model/datacatalog/api.go [hunk 8]
```diff
@@ -42,3 +42,15 @@ type CreateAssetResponse struct {
...
+type DeleteAssetRequest struct {
+	// The destination catalog id in which the asset to be deleted exists
+	DestinationCatalogID string `json:"destinationCatalogID"`
+	// Asset ID of the to-be deleted asset
+	DestinationAssetID string `json:"destinationAssetID,omitempty"`
+}
+
+type DeleteAssetResponse struct {
+	// The ID of the created asset based on the source asset information given in CreateAssetRequest object
+	Status string `json:"status"`
```
Comment (shlomitk1) [inline pkg/model/datacatalog/api.go:55]: See an earlier comment about the status

File: manager/controllers/mockup/datacatalog.go (hunk 5 above)
Comment (shlomitk1) [reply manager/controllers/mockup/datacatalog.go:60]: BTW, where is the function definition in the katalog-connector?

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 4 above)
Comment (roytman) [reply charts/fybrik/files/taxonomy/datacatalog.json:65]: Yes, I'd change it to an optional message

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go [hunk 9]
```diff
@@ -106,35 +107,37 @@ func ConvertDataCatalogGrpcRespToOpenAPIResp(result *pb.CatalogDatasetInfo) (*da
...
-			// just printing - start
-			responseBytes, errJSON = json.MarshalIndent(&rscCol, "", "\t")
-			if errJSON != nil {
-				return nil, fmt.Errorf("error Marshalling in ConvertDataCatalogGrpcRespToOpenAPIResp: %v", errJSON)
-			}
-			log.Print("responseBytes after MarshalIndent in ConvertDataCatalogGrpcRespToOpenAPIResp:" + string(responseBytes))
-			// just printing - end
-
-			resourceCols = append(resourceCols, rscCol)
+		if compMetaData == nil {
```
Comment (roytman) [inline pkg/connectors/datacatalog/clients/datacatalog_grpc.go:110]: how do all these changes relate to the PR?

File: pkg/model/datacatalog/api.go [hunk 10]
```diff
@@ -42,3 +42,15 @@ type CreateAssetResponse struct {
 	// The ID of the created asset based on the source asset information given in CreateAssetRequest object
 	AssetID string `json:"assetID"`
 }
+
+type DeleteAssetRequest struct {
+	// The destination catalog id in which the asset to be deleted exists
+	DestinationCatalogID string `json:"destinationCatalogID"`
```
Comment (roytman) [inline pkg/model/datacatalog/api.go:48]: again, why not `AssetID taxonomy.AssetID` ? 
@shlomitk1  ?

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 4 above)
Comment (rohithdv) [reply charts/fybrik/files/taxonomy/datacatalog.json:65]: Thanks. Could you give some pointers to the error codes we use for delete operation? I can provide a required error code field in the `DeleteAssetResponse` object while making `Status` field optional. @shlomitk1

File: site/docs/reference/connectors-datacatalog/Apis/DefaultApi.md [hunk 11]
```diff
@@ -5,6 +5,7 @@ All URIs are relative to *https://localhost:8080*
 Method | HTTP request | Description
 ------------- | ------------- | -------------
 [**createAsset**](DefaultApi.md#createAsset) | **POST** /createAsset | This REST API writes data asset information to the data catalog configured in fybrik
+[**deleteAsset**](DefaultApi.md#deleteAsset) | **POST** /deleteAsset | This REST API deletes data asset from the data catalog configured in fybrik
```
Comment (roytman) [inline site/docs/reference/connectors-datacatalog/Apis/DefaultApi.md:8]: I'd change it to just " **deletes** data asset ... " it is relevant to all calls

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go (hunk 9 above)
Comment (rohithdv) [reply pkg/connectors/datacatalog/clients/datacatalog_grpc.go:110]: This was fixed as `golangci-lint` was complaining about this error.

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 4 above)
Comment (roytman) [reply charts/fybrik/files/taxonomy/datacatalog.json:65]: @rohithdv , see https://www.restapitutorial.com/lessons/httpmethods.html

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 1 above)
Comment (shlomitk1) [reply charts/fybrik/files/taxonomy/datacatalog.json:50]: Yes, it is exactly the same one. No need for destinationCatalogID.

File: pkg/model/datacatalog/api.go (hunk 10 above)
Comment (shlomitk1) [reply pkg/model/datacatalog/api.go:48]: Yes, should be `AssetID taxonomy.AssetID `json:"assetID"`` as in GetAssetRequest.

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 4 above)
Comment (rohithdv) [reply charts/fybrik/files/taxonomy/datacatalog.json:65]: @roytman I had one query. These status codes will be anyways being sent from the `deleteAsset` implementation. So we need not repeat these codes in the response object right ? I am not sure what should be provided in the `DeleteAssetResponse` object. Could you elaborate on that? @shlomitk1

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go (hunk 6 above)
Comment (rohithdv) [reply pkg/connectors/datacatalog/clients/datacatalog_grpc.go:13]: First we need to list where all we have used grpc. Probably we can raise a git issue first enlisting the various parts of the source which needs to be removed. Then remove it in a PR. Does this sound ok? @shlomitk1

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go (hunk 7 above)
Comment (rohithdv) [reply pkg/connectors/datacatalog/clients/datacatalog_grpc.go:220]: We need to do this as we are enforcing the data catalog interface on the grpc implementation (https://github.com/fybrik/fybrik/blob/451c4192920c132350d83b1678edbff3f34c90a7/pkg/connectors/datacatalog/clients/datacatalog_grpc.go#L250).

File: manager/controllers/mockup/datacatalog.go (hunk 5 above)
Comment (rohithdv) [reply manager/controllers/mockup/datacatalog.go:60]: In the next PR, we will provide the implementation for katalog-connector.

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 1 above)
Comment (rohithdv) [reply charts/fybrik/files/taxonomy/datacatalog.json:50]: Fixed. Thanks.

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 2 above)
Comment (rohithdv) [reply charts/fybrik/files/taxonomy/datacatalog.json]: Fixed. Thanks.

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 3 above)
Comment (rohithdv) [reply charts/fybrik/files/taxonomy/datacatalog.json]: Fixed. Thanks.

File: pkg/model/datacatalog/api.go (hunk 8 above)
Comment (rohithdv) [reply pkg/model/datacatalog/api.go]: Fixed. Thanks.

File: pkg/connectors/datacatalog/clients/datacatalog_openapi.go [hunk 12]
```diff
@@ -60,6 +60,14 @@ func (m *openAPIDataCatalog) CreateAsset(in *datacatalog.CreateAssetRequest, cre
 	return resp, nil
 }
 
+func (m *openAPIDataCatalog) DeleteAsset(in *datacatalog.DeleteAssetRequest, creds string) (*datacatalog.DeleteAssetResponse, error) {
+	resp, _, err := m.client.DefaultApi.DeleteAsset(context.Background()).DeleteAssetRequest(*in).Execute()
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply pkg/connectors/datacatalog/clients/datacatalog_openapi.go]: Regarding first point, response object which is returned from `m.client.DefaultApi.DeleteAsset` call will have the optional error message. Did I get your question right?

[the remark this answers is not in the document]
Comment (rohithdv) [reply pkg/connectors/datacatalog/clients/datacatalog_openapi.go]: Provided the fix for the second point. Thanks.

File: pkg/model/datacatalog/api.go (hunk 10 above)
Comment (rohithdv) [reply pkg/model/datacatalog/api.go]: Fixed. Thanks.

File: site/docs/reference/connectors-datacatalog/Apis/DefaultApi.md (hunk 11 above)
Comment (rohithdv) [reply site/docs/reference/connectors-datacatalog/Apis/DefaultApi.md]: Fixed. Thanks.

File: charts/fybrik/files/taxonomy/datacatalog.json (hunk 4 above)
Comment (shlomitk1) [reply charts/fybrik/files/taxonomy/datacatalog.json]: An optional error message

**Review by roytman — APPROVED** [review]
